### ISSUE TITLE
Issue with pblapply Parallel Execution Efficiency: Non-Continuous Parallel Execution Leading to Low Resource Utilization

### ISSUE THREAD

**Issue by daydayfei** [psolymos/pbapply#70]
grouped_list <- mydata %>%  
      group_by(id.exposure, id.outcome) %>%  
      group_split()
    
    group_sizes <- sapply(grouped_list, nrow)  
    
    sorted_grouped_list <- grouped_list[order(group_sizes, decreasing = TRUE)]  
 res =  pblapply(X =sorted_grouped_list, presso_function, cl = cl)

Problem Description:

When using the pblapply function (assuming it's from the parallel package or a similar parallel processing package) for parallel computing, I've encountered an issue regarding the efficiency of parallel execution. Specifically, despite setting up multiple worker threads (e.g., specifying 10 worker threads using the cl parameter), pblapply does not continuously parallelize the processing of all tasks as I expected. Instead, it seems to run tasks in batches, waiting for the slowest task in each batch to complete before starting the next batch, rather than keeping all worker threads busy at all times.

Expected Behavior:

I expect pblapply to utilize all available worker threads to continuously and in parallel process each task in the task list, regardless of the differences in execution time between tasks. This way, even if some tasks take longer than others, other worker threads can continue processing other pending tasks, improving overall execution efficiency.

Actual Behavior:

In actual usage, I observe that pblapply tends to wait for the slowest task in the current batch (assuming 10 tasks per batch) to complete before starting the next batch. This execution mode leads to low resource utilization (CPU cores and memory), particularly when there are significant differences in task execution times.

Example Scenario:

Suppose we have a task list of 100 tasks, with varying computational loads. I use 10 worker threads to execute these tasks. Ideally, all 10 worker threads should start working almost simultaneously, and as soon as a task completes, a new task is immediately assigned to an idle worker thread. However, in reality, I see that worker threads wait for all tasks in the first batch to complete before processing the next batch, and so on.

Impact:

This non-continuous parallel execution pattern significantly reduces overall computational efficiency, especially when dealing with large-scale datasets and computationally intensive tasks. It increases computation time and may lead to unnecessary resource idling.

Request:

I request that the developers consider optimizing the parallel execution mechanism of pblapply to ensure that all worker threads can continuously and efficiently process tasks in parallel, without waiting for the slowest task in a batch to complete. This may involve improving the task scheduling algorithm or adding dynamic management of task execution progress.

**Comment by psolymos** [comment]
Hi @daydayfei take a look at some of the options (`?pboptions`), especially choosing a smaller `nout` value and turning load balancing on with `use_lb=TRUE`. You can also take a look at using the future backend with `cl = "future"`.
